File: python/elf_parser.py
import struct


class ElfHeader:
    def __init__(self, file):
        self.e_ident = {}

        e_ident_restrictions = [
            ['MAG0', [0x7F]],
            ['MAG1', [0x45]],
            ['MAG2', [0x4C]],
            ['MAG3', [0x46]],
            ['EI_CLASS', [1, 2]],
            ['EI_DATA', [1, 2]],
            ['EI_VERSION', [1]],
            ['EI_OSABI', [*range(0x12+1)]],
            ['EI_ABIVERSION', [*range(256)]],
            ['EI_PAD0', [0]],
            ['EI_PAD1', [0]],
            ['EI_PAD2', [0]],
            ['EI_PAD3', [0]],
            ['EI_PAD4', [0]],
            ['EI_PAD5', [0]],
            ['EI_PAD6', [0]],
        ]

        for field in e_ident_restrictions:
            name = field[0]
            value_range = field[1]
            value = list(file.read(1))[0]
            self.e_ident[name] = value

            if value not in value_range:
                raise Exception("bad %s in ELF header: 0x%02x" % [name, value])

        scheme = "<" if self.is_little_endian() else ">"
        if self.is_64_bit():
            scheme += 'HHLQQQLHHHHHH'
        else:
            scheme += 'HHLLLLLHHHHHH'

        buffer = file.read(struct.calcsize(scheme))

        self.e_type, self.e_machine, self.e_version, self.e_entry,
        self.e_phoff, self.e_shoff, self.e_flags, self.e_ehsize,
        self.e_phentsize, self.e_phnum, self.e_shentsize,
        self.e_shnum, self.e_shstrndx = struct.unpack(scheme, buffer)

    def is_little_endian(self):
        return self.e_ident['EI_DATA'] == 1

    def is_64_bit(self):
        return self.e_ident['EI_CLASS'] == 2


def read_data_to_load_to_memory(file, offset, size_in_file, size_in_memory):
    debug_info = "offset: %d" % offset
    debug_info += ", size in file: %d" % size_in_file
    debug_info += ", size in memory: %d" % size_in_memory
    print(debug_info)

    file.seek(offset)
    bytes = list(file.read(size_in_file))

    # fill with zeros if size_in_file < size_in_memory
    bytes += [0] * size_in_memory

    # take only the part that fits into memory
    return bytes[:size_in_memory]


class ElfProgramHeaderEntry:
    def __init__(self, file, elf_header):

        scheme = "<" if elf_header.is_little_endian() else ">"
        if elf_header.is_64_bit():
            scheme += "LLQQQQQQ"

            buffer = file.read(struct.calcsize(scheme))

            self.p_type, self.p_flags, self.p_offset, self.p_vaddr,\
                self.p_paddr, self.p_filesz, self.p_memsz, self.p_align =\
                struct.unpack(scheme, buffer)
        else:
            scheme += "LLLLLLLL"

            buffer = file.read(struct.calcsize(scheme))

            self.p_type, self.p_offset, self.p_vaddr, self.p_paddr,\
                self.p_filesz, self.p_memsz, self.p_flags, self.p_align =\
                struct.unpack(scheme, buffer)

        self.data_to_load_to_memory = \
            read_data_to_load_to_memory(file, self.p_offset, self.p_filesz,
                                        self.p_memsz)


class ElfProgramHeader:
    def __init__(self, file, elf_header):
        self.entries = []
        for i in range(elf_header.e_phnum):
            file.seek(elf_header.e_phoff + elf_header.e_phentsize * i)
            self.entries.append(ElfProgramHeaderEntry(file, elf_header))


class ElfSectionHeaderEntry:
    def __init__(self, file, elf_header):

        scheme = "<" if elf_header.is_little_endian else ">"
        if elf_header.is_64_bit():
            scheme += "LLQQQQLLQQ"
        else:
            scheme += "LLLLLLLLLL"

        buffer = file.read(elf_header.e_shentsize)

        self.sh_name, self.sh_type, self.sh_flags, self.sh_addr,
        self.sh_offset, self.sh_size, self.sh_link, self.sh_info,
        self.sh_addralign, self.sh_entsize = struct.unpack(scheme, buffer)


class ElfSectionHeader:
    def __init__(self, file, elf_header):
        self.entries = []
        for i in range(elf_header.e_shnum):
            file.seek(elf_header.e_shoff + i * elf_header.e_shentsize)
            self.entries.append(ElfSectionHeaderEntry(file, elf_header))


class Elf:
    def __init__(self, filename):
        file = open(filename, "rb")

        self.elf_header = ElfHeader(file)
        self.program_header = ElfProgramHeader(file, self.elf_header)
        self.section_header = ElfSectionHeader(file, self.elf_header)

File: elf_parser.rb
class ElfHeader
    def initialize(file)
        @e_ident = {}
        
        e_ident_restrictions = [
            ['MAG0', [0x7F]],
            ['MAG1', [0x45]],
            ['MAG2', [0x4C]],
            ['MAG3', [0x46]],
            ['EI_CLASS', [1, 2]],
            ['EI_DATA', [1, 2]],
            ['EI_VERSION', [1]],
            ['EI_OSABI', (0..0x12).to_a],
            ['EI_ABIVERSION', (0..0xFF).to_a],
            ['EI_PAD0', [0]],
            ['EI_PAD1', [0]],
            ['EI_PAD2', [0]],
            ['EI_PAD3', [0]],
            ['EI_PAD4', [0]],
            ['EI_PAD5', [0]],
            ['EI_PAD6', [0]],
        ]
        
        e_ident_restrictions.each do |field|
            name = field[0]
            value_range = field[1]
            value = file.read(1).unpack('C')[0]
            @e_ident[name] = value
            raise "bad %s in ELF header: 0x%02x" % [name, value] unless value_range.include? value
        end
        
        scheme = is_little_endian ? (is_64_bit ? 'S<S<L<Q<Q<Q<L<S<S<S<S<S<S<' : 'S<S<L<L<L<L<L<S<S<S<S<S<S<')
                                  : (is_64_bit ? 'S>S>L>Q>Q>Q>L>S>S>S>S>S>S>' : 'S>S>L>L>L>L>L>S>S>S>S>S>S>')
        
        remaining_header_size = (is_64_bit ? 0x40 : 0x34) - 0x10
        
        @e_type, @e_machine, @e_version, @e_entry, @e_phoff,
        @e_shoff, @e_flags, @e_ehsize, @e_phentsize, @e_phnum,
        @e_shentsize, @e_shnum, @e_shstrndx = file.read(remaining_header_size).unpack(scheme)
    end

    attr_reader :e_type, :e_machine, :e_version, :e_entry, :e_phoff,
                :e_shoff, :e_flags, :e_ehsize, :e_phentsize, :e_phnum,
                :e_shentsize, :e_shnum, :e_shstrndx
    
    def is_little_endian
        return @e_ident['EI_DATA'] == 1
    end
    
    def is_64_bit
        return @e_ident['EI_CLASS'] == 2
    end
end

def read_data_to_load_to_memory(file, offset, size_in_file, size_in_memory)
    puts "offset: %d, size in file: %d, size in memory: %d" % [offset, size_in_file, size_in_memory]
    file.seek(offset)
    bytes = file.read(size_in_file).unpack("C*")
    # fill with zeros if size_in_file < size_in_memory
    bytes += Array.new(size_in_memory, 0)
    # take only the part that fits into memory
    return bytes.slice(0, size_in_memory)
end

class ElfProgramHeaderEntry
    def initialize(file, elf_header)
        if elf_header.is_64_bit
            scheme = elf_header.is_little_endian ? "L<L<Q<Q<Q<Q<Q<Q<" : "L>L>Q>Q>Q>Q>Q>Q>";
            @p_type, @p_flags, @p_offset, @p_vaddr,
            @p_paddr, @p_filesz, @p_memsz, @p_align = file.read(elf_header.e_phentsize).unpack(scheme)
        else
            scheme = elf_header.is_little_endian ? "L<L<L<L<L<L<L<L<" : "L>L>L>L>L>L>L>L>";
            @p_type, @p_offset, @p_vaddr, @p_paddr,
            @p_filesz, @p_memsz, @p_flags, @p_align = file.read(elf_header.e_phentsize).unpack(scheme)
        end
        
        @data_to_load_to_memory = read_data_to_load_to_memory(file, @p_offset, @p_filesz, @p_memsz)
    end
    
    attr_reader :p_type, :p_flags, :p_offset, :p_vaddr,
                :p_paddr, :p_filesz, :p_memsz, :p_align,
                :data_to_load_to_memory
end

class ElfProgramHeader
    def initialize(file, elf_header)
        @entries = []
        for i in 0..elf_header.e_phnum - 1 do
            file.seek(elf_header.e_phoff + elf_header.e_phentsize * i)
            @entries.append(ElfProgramHeaderEntry.new(file, elf_header))
        end
    end
    
    attr_reader :entries
end

class ElfSectionHeaderEntry
    def initialize(file, elf_header)
        if elf_header.is_64_bit
            scheme = elf_header.is_little_endian ? "L<L<Q<Q<Q<Q<L<L<Q<Q<" : "L>L>Q>Q>Q>Q>L>L>Q>Q>"
        else
            scheme = elf_header.is_little_endian ? "L<L<L<L<L<L<L<L<L<L<" : "L>L>L>L>L>L>L>L>L>L>"
        end
        @sh_name, @sh_type, @sh_flags, @sh_addr, @sh_offset, @sh_size, @sh_link,
        @sh_info, @sh_addralign, @sh_entsize = file.read(elf_header.e_shentsize).unpack(scheme)
    end
    
    attr_reader :sh_name, :sh_type, :sh_flags, :sh_addr, :sh_offset, :sh_size, :sh_link,
                :sh_info, :sh_addralign, :sh_entsize
end

class ElfSectionHeader
    def initialize(file, elf_header)
        @entries = []
        for i in 0..elf_header.e_shnum - 1 do
            file.seek(elf_header.e_shoff + i * elf_header.e_shentsize)
            @entries.append(ElfSectionHeaderEntry.new(file, elf_header))
        end
    end
    
    attr_reader :entries
end

class Elf
    def initialize(filename)
        File.open(filename) do |file|
            @elf_header = ElfHeader.new(file)
            @program_header = ElfProgramHeader.new(file, @elf_header)
            @section_header = ElfSectionHeader.new(file, @elf_header)
        end
    end
    
    attr_reader :elf_header, :program_header, :section_header
end